pkg/sidecar-controller/groupsnapshot_helper.go

```diff
@@ -0,0 +1,668 @@
+// * eventtype, reason, message - event to send, see EventRecorder.Event()
+func (ctrl *csiSnapshotSideCarController) updateGroupSnapshotContentErrorStatusWithEvent(groupSnapshotContent *crdv1alpha1.VolumeGroupSnapshotContent, eventtype, reason, message string) error {
+	klog.V(5).Infof("updateGroupSnapshotContentStatusWithEvent[%s]", groupSnapshotContent.Name)
+
+	if groupSnapshotContent.Status != nil && groupSnapshotContent.Status.Error != nil && *groupSnapshotContent.Status.Error.Message == message {
+		klog.V(4).Infof("updateGroupSnapshotContentStatusWithEvent[%s]: the same error %v is already set", groupSnapshotContent.Name, groupSnapshotContent.Status.Error)
+		return nil
+	}
+
+	var patches []utils.PatchOp
+	ready := false
+	groupSnapshotContentStatusError := &crdv1.VolumeSnapshotError{
+		Time: &metav1.Time{
+			Time: time.Now(),
+		},
+		Message: &message,
+	}
+	if groupSnapshotContent.Status == nil {
+		// Initialize status if nil
+		patches = append(patches, utils.PatchOp{
+			Op:   "replace",
+			Path: "/status",
+			Value: &crdv1alpha1.VolumeGroupSnapshotContentStatus{
+				ReadyToUse: &ready,
+				Error:      groupSnapshotContentStatusError,
+			},
+		})
+	} else {
+		// Patch status if non-nil
+		patches = append(patches, utils.PatchOp{
+			Op:    "replace",
+			Path:  "/status/error",
+			Value: groupSnapshotContentStatusError,
+		})
+		patches = append(patches, utils.PatchOp{
+			Op:    "replace",
+			Path:  "/status/readyToUse",
+			Value: &ready,
+		})
+
+	}
+
+	newContent, err := utils.PatchVolumeGroupSnapshotContent(groupSnapshotContent, patches, ctrl.clientset, "status")
+
+	// Emit the event even if the status update fails so that user can see the error
+	ctrl.eventRecorder.Event(newContent, eventtype, reason, message)
+
+	if err != nil {
+		klog.V(4).Infof("updating VolumeGroupSnapshotContent[%s] error status failed %v", groupSnapshotContent.Name, err)
+		return err
+	}
+
+	_, err = ctrl.storeGroupSnapshotContentUpdate(newContent)
+	if err != nil {
+		klog.V(4).Infof("updating VolumeGroupSnapshotContent[%s] error status: cannot update internal cache %v", groupSnapshotContent.Name, err)
+		return err
+	}
+
+	return nil
+}
+
+// GetSnapshotNameForVolumeGroupSnapshotContent returns a unique snapshot name for a VolumeGroupSnapshotContent.
+func GetSnapshotNameForVolumeGroupSnapshotContent(groupSnapshotContent *crdv1alpha1.VolumeGroupSnapshotContent) string {
+	return fmt.Sprintf("groupsnapshot-%x-%d", sha256.Sum256([]byte(groupSnapshotContent.UID)), time.Duration(time.Now().UnixNano())/time.Millisecond)
+}
+
+// GetSnapshotContentNameForVolumeGroupSnapshotContent returns a unique content name for the
+// passed in VolumeGroupSnapshotContent.
+func GetSnapshotContentNameForVolumeGroupSnapshotContent(groupSnapshotContent *crdv1alpha1.VolumeGroupSnapshotContent) string {
+	return fmt.Sprintf("groupsnapcontent-%x-%d", sha256.Sum256([]byte(groupSnapshotContent.UID)), time.Duration(time.Now().UnixNano())/time.Millisecond)
```

Comment:
To be consistent with the naming of regular VolumeSnapshotContent, let's make the prefix "snapcontent-".

Reply:
Done